code/assurance/src/main/java/com/adobe/marketing/mobile/assurance/AssuranceSessionOrchestrator.java

```diff
@@ -76,20 +76,61 @@ public void onConnect(final String pin) {
                                 Assurance.LOG_TAG,
                                 LOG_TAG,
                                 "Null/Empty PIN recorded. Cannot connect to session.");
-                        terminateSession();
+                        terminateSession(true);
                         return;
                     }
 
                     session.connect(pin);
                 }
 
+                @Override
+                public void onQuickConnect(
+                        @NonNull String sessionId,
+                        @NonNull String token,
+                        @NonNull AssuranceSession.AssuranceSessionStatusListener listener) {
+
+                    if (session != null) {
+                        // There is an active session,  check the authorizing presentation to
+                        // determine if a new session can be started or not.
+                        if (session.getAuthorizingPresentationType()
+                                == SessionAuthorizingPresentation.Type.PIN) {
+                            // This should never happen in an ideal situation.
+                            Log.warning(
+                                    Assurance.LOG_TAG,
+                                    LOG_TAG,
+                                    "Cannot start Quick Connect session. An active PIN based"
+                                            + " session exists.");
+                            listener.onSessionTerminated(
+                                    AssuranceConstants.AssuranceConnectionError.UNEXPECTED_ERROR);
+                            return;
+                        } else {
+                            // This is a Quick Connect retry scenario. Disconnect existing session
+                            // without clearing the buffered events for this retry scenario and
+                            // connect again.
+                            Log.debug(
+                                    Assurance.LOG_TAG,
+                                    LOG_TAG,
+                                    "Disconnecting active QuickConnect session and recreating.");
+
+                            terminateSession(false);
```

Comment:
Can you explain this path further? I am confused about when we would want to terminate an existing session to create a new one? On iOS we just log a warning when someone attempts to create a new session when a session exists. Want to know if we need to rethink this on iOS or on Android.

Reply:
There are 2 cases when a request to create a new session when an active session exists (either connected or disconnected but retryable) can happen:

- First, it can be via triggering the `startSession()` api when there is already an existing session. In this case we should _always_ bail and log a warning. That piece of code is the same in Android and is coming up after this PR 😄 

- Second, and the one this specific path tackles is the case where:
   1. QuickConnect flow is first initiated via the `startSession()` public API that sends an event to the Assurance extension to start a session. 
   2. Quick connect screen is displayed and the user clicked "Connect"
  3. Assurance attempts to create a session and connect to the session but failed. Now a "Retry" button is being shown.
  4. We implemented in iOS and Android retry logic in a way that we restart the entire creation+connection process when the quick connect session creation fails. So we should terminate the existing session when the user clicks "Retry" and the copy over the buffered events to create a new session.
  Ideally in the next revision we should be retrying particular phases intelligently based on the nature of error (creation phase error , connection phase error) during quick connect.

Let me know if this clarifies or we can hop over a call to discuss further.

Reply:
So on Android, you create a session before the quick connect flow is complete?

Reply:
Yes, after the user clicks on the connect button on the QuickConnect flow - the session is created once we have the sessionId and token from the device API calls . At this time, if a session connection is successful, then we dismiss the quick connect screen, if the connection fails due to an error - the error is shown on the screen with a retry button per the spec.

Reply:
I was misunderstanding something because of how we do it on iOS (i've found a bug on iOS because of this, thanks! 🤣)